### PR TITLE
Implement Recovery compass

### PR DESCRIPTION
**Pull request by IvanCraft623**
## Introduction
Implements recovery compass, this item points to the position of the player's last kill, the position is stored with the player data and send via entity properties.

### API changes

The following methods has been added to the `Player` class:
- `Player:getDeathPosition() : ?Position`
- `Player:setDeathPosition(?Position $position) : void`

## Follow up
- Update BedrockPtotocol!!!

## Tests
https://youtu.be/cdTEHqx5c7U

### CODE REVIEW

File: src/player/Player.php [hunk 1]
```diff
@@ -2248,6 +2280,15 @@ public function getSaveData() : CompoundTag{
 			$nbt->setInt(self::TAG_SPAWN_Z, $spawn->getFloorZ());
 		}
 
+		if($this->deathPosition !== null){
+			$nbt->setString(self::TAG_DEATH_WORLD, $this->deathPosition->getWorld()->getFolderName());
+			$nbt->setInt(self::TAG_DEATH_X, $this->deathPosition->getFloorX());
+			$nbt->setInt(self::TAG_DEATH_Y, $this->deathPosition->getFloorY());
+			$nbt->setInt(self::TAG_DEATH_Z, $this->deathPosition->getFloorZ());
+		}else{
+			$nbt->removeTag(self::TAG_DEATH_WORLD, self::TAG_DEATH_X, self::TAG_DEATH_Y, self::TAG_DEATH_Z);
```
Comment (dktapps) [inline src/player/Player.php]: This is redundant. The NBT is freshly created, so it won't have these tags anyway.

File: src/player/Player.php [hunk 2]
```diff
@@ -2248,6 +2280,15 @@ public function getSaveData() : CompoundTag{
 			$nbt->setInt(self::TAG_SPAWN_Z, $spawn->getFloorZ());
 		}
 
+		if($this->deathPosition !== null){
+			$nbt->setString(self::TAG_DEATH_WORLD, $this->deathPosition->getWorld()->getFolderName());
```
Comment (dktapps) [inline src/player/Player.php:2416]: This may throw an exception if the world was unloaded since the player died

File: src/player/Player.php [hunk 3]
```diff
@@ -969,6 +979,27 @@ public function doChunkRequests() : void{
...
+	public function getDeathPosition() : ?Position{
+		return $this->deathPosition;
+	}
+
+	/**
+	 * @param Vector3|Position|null $pos
+	 */
+	public function setDeathPosition(?Vector3 $pos) : void{
+		if($pos !== null){
+			if($pos instanceof Position && $pos->world !== null){
```
Comment (dktapps) [inline src/player/Player.php:1057]: I think it's fine to just assert that the position must be valid here. We shouldn't be setting death position to an unloaded world anyway.

**Review by dktapps — CHANGES_REQUESTED** [review]

File: src/player/Player.php (hunk 2 above)
Comment (IvanCraft623) [reply src/player/Player.php:2416]: Would you agree to store the world's name in a var on `Player::setDeathPosition()`?

Comment (dktapps) [reply src/player/Player.php:2416]: It should probably be done in a similar way to spawn position handling, where the spawn is invalidated if the world isn't loaded. I'm not sure this is the _best_ way, but it should be consistent.

File: src/player/Player.php (hunk 3 above)
Comment (dktapps) [reply src/player/Player.php:1057]: Apparently this is the same as how the spawn setting logic works. Perhaps we need to take a hard look at why Position is allowed to have a null `World`...

**Review by dktapps — APPROVED** [review]

**Pull request merged**